### DISCUSSION TITLE
How to return values from a CI-independent cake script?

### DISCUSSION THREAD

**Discussion by bergziege** [cake-build/cake#3102] [Q&A] [answered]
Hi, I am currently having the use case where I whould like write a cake which does not have any CI tool specific code in it but does the CI specific steps afterwards (or beforehand).

Example: Deploy "App-1.2.3-beta-4.zip" (not a static artifact) to the azure devops pipeline artifacts

The question is: How do I pass value/values from Cake back to the calling pipeline?

Reasons: To be able to quickly migrate the script between gitlab/azure/jenkins without the need to integrate all those informations into the script itself.

Could not find a non CI tool specific way to do so far (eg. setting a env var from within cake).

**Comment by pascalberger** [comment, marked as answer]
The problem with environment variables is that they are set for the running process. Every CI system has a slightly different and specific approach to handle this scenario.
What I do in these cases is to write an interface as abstraction layer, with concrete implementations for different CI systems.

[Cake.Issues.Recipe](https://github.com/cake-contrib/Cake.Issues.Recipe/tree/develop/Cake.Issues.Recipe/Content/tasks/buildservers) is an example for this.

**Reply by bergziege** [reply]
Thanks for the tip, but writing concrete implementations was exactly what I was hoping to avoid. I had something like "return a magic json string" in mind.
But maybe I should simply go the old convention based way and let cake prepare some folders where everything is stored for the next process.
